subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -173,27 +191,42 @@ private String getFilename(Path dir) {
 
         @Override
         public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
-            String name = stringInterner.intern(file.getFileName().toString());
-            if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
+            String name = intern(file.getFileName().toString());
+            if (shouldVisit(file, name, false, attrs, builder.getRelativePath())) {
                 if (attrs == null) {
                     throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
                 }
-                if (attrs.isSymbolicLink()) {
-                    // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
-                    throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
-                }
-                addFileSnapshot(file, name, attrs);
+                builder.visit(snapshot(file, name, attrs));
             }
             return FileVisitResult.CONTINUE;
         }
 
+        private FileSystemLocationSnapshot snapshot(Path absoluteFilePath, String name, BasicFileAttributes attrs) {
+            String internedAbsoluteFilePath = intern(absoluteFilePath.toString());
+            if (attrs.isRegularFile()) {
+                try {
+                    HashCode hash = hasher.hash(absoluteFilePath.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
+                    FileMetadata metadata = FileMetadata.from(attrs);
+                    return new RegularFileSnapshot(internedAbsoluteFilePath, name, hash, metadata);
+                } catch (UncheckedIOException e) {
+                    LOGGER.debug("Could not read file path '{}'.", absoluteFilePath, e);
+                }
+            }
+            return new UnavailableFileSnapshot(internedAbsoluteFilePath, name);
+        }
+
         @Override
         public FileVisitResult visitFileFailed(Path file, IOException exc) {
             // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
             // so we include all the other files apart from the loop.
             // This way, we include each file only once.
-            if (isNotFileSystemLoopException(exc) && isAllowed(file, file.getFileName().toString(), false, null, builder.getRelativePath())) {
-                throw new UncheckedIOException(String.format("Could not read path '%s'.", file), exc);
+            if (isNotFileSystemLoopException(exc)) {
+                String name = intern(file.getFileName().toString());
+                if (shouldVisit(file, name, Files.isDirectory(file), null, builder.getRelativePath())) {
```

Comment:
I would suggest we use `false` here. I looks like `isDirectory` uses `readAttributes`, which will return `null`, and then `isDirectory` returns `false`. Is there another possible path here?

Reply:
Running the new unreadable copy test actually returns `true` for `Files.isDirectory(file)` here in both cases (files/folders)

Reply:
It looks like `visitFailed` is only called for unreadable directories and not for unreadable files. That is why `isDirectory(file) == true` I guess.

Reply:
yes, it seems to me like that, but it was previously a constant `false`. Was that a bug?

Reply:
I don't think it was necessarily a bug. See https://github.com/gradle/gradle/pull/9746/files#r299094146.